File: Python/Graphs/L_pathWithMaxProb.py
import collections
import heapq
from typing import List

def maxProbability(self, n: int, edges: List[List[int]], succProb: List[float], start: int, end: int) -> float:
        adj = collections.defaultdict(list)
        for i in range(len(edges)):
            src, dest = edges[i]
            prob = succProb[i]
            adj[src].append([dest, prob])
            adj[dest].append([src, prob])
            
        pq = [(-1, start)]
        visited = set()
        
        while pq:
            prob, curr = heapq.heappop(pq)
            visited.add(curr)
            
            if curr == end:
                return prob * -1
            
            for nei, edgeProb in adj[curr]:
                if nei not in visited:
                    heapq.heappush(pq, ((prob * edgeProb), nei))
        
        return 0

File: JavaScript/Graph/L_pathWithMaxProb.js
var maxProbability = function (n, edges, succProb, start_node, end_node) {
    let maxProb = new Array(n).fill(0.0);
    maxProb[start_node] = 1.0;

    for (let i = 0; i < n - 1; i++) {
        let updated = false;
        for (let j = 0; j < edges.length; j++) {
            let u = edges[j][0];
            let v = edges[j][1];
            let prob = succProb[j];

            if (maxProb[u] * prob > maxProb[v]) {
                maxProb[v] = maxProb[u] * prob;
                updated = true;
            }
            if (maxProb[v] * prob > maxProb[u]) {
                maxProb[u] = maxProb[v] * prob;
                updated = true;
            }
        }
        if (!updated) break;
    }

    return maxProb[end_node];
};
